designs/s3/20200210-design-vscode-s3.md

```diff
@@ -0,0 +1,146 @@
+# UX: S3 explorer
+
+* **Type**: UX Design
+* **Status**: Proposal
+
+## Overview
+
+This document proposes a user experience for interacting with AWS S3 from the
+AWS Toolkit ("the Toolkit") for Visual Studio Code ("VSCode").
+
+Some of the features listed below are vague; they are implemented using VSCode
+or platform conventions analogous to the AWS web console.
+
+## Experience
+
+Each region of the Toolkit explorer tree includes an S3 node (see mockup
+below). The toolkit pulls a list of names and properties of S3 objects in the
+currently active profile and presents this data as a tree in the toolkit
+explorer. This is typical of existing toolkit features such as "Schemas" and
+"CDK".
+
+- When a region is expanded, make an eager request to get the count of S3
+  buckets for that region.
+- Show the count of buckets next to the S3 node.
+- The S3 node is always present. If there are no buckets in the region, "grey
+  out" the S3 node.
+- When the current active profile changes, the tree is not refreshed, but is
+  greyed-out to indicate that it is stale. It is not minimized.
+- When a bucket or object is created or deleted, only that node is added or
+  removed from the tree, _without disturbing the current layout of the tree_.
+
+
+![](20200210-design-vscode-s3-tree.png)
+
+### Create and edit S3 buckets
+
+- Root node context menu (right-click):
+  - Create bucket
+- Bucket node context menu (right-click):
+  - Delete bucket
+  - New folder...
+  - Upload...
+  - Copy name
+  - Copy ARN
+
+### Create and edit S3 objects
+
+- S3 object node context menu (right-click):
+  - Delete object
+  - Copy name
+  - Copy ARN
+  - Copy URL
+  - Upload...
```

Comment:
> What does Upload on an object node do

It uploads to the folder. This avoids "traps" where the user is punished for having a non-folder item selected, which then requires the user to carefully select the parent folder--even though it is obvious where the user wants to upload to.

Reply:
Just as a note, the current VS Code explorer (not the toolkit) does not work like that, so it may be a little inconsistent with the IDE experience.  They get around these "traps"(sort of) by adding pop-up options for common tasks like below.  Choosing "New Folder" or "New File" will add the new element into the active folder. 
![vscodeExplorerAddFile](https://user-images.githubusercontent.com/25124281/75296413-f6514200-57e1-11ea-9c50-21c83543e27c.png)

Reply:
> VS Code explorer (not the toolkit) does not work like that

You mean the items aren't shown in the context menu. But the "active folder" is not affected by whatever file happens to be highlighted, and clicking "New folder" or "New file" choose the parent of the highlighted file.

Reply:
Upload menu item on file
> even though it is obvious where the user wants to upload to

If it was obvious to me, I wouldn't have asked what this is supposed to do. I could see this doing either one of the following actions:
* replace selected object with uploaded contents
* upload into parent folder

These are rather different outcomes. 

I like @gsweetwood 's proposal for its consistency, except that the "new file" flow might have to take a slightly different approach than "create a empty object with a new file name"

Reply:
+1 for being consistent with the file-explorer VS Code already has - this is the idiomatic experience.

Reply:
> This avoids "traps" where the user is punished for having a non-folder item selected, which then requires the user to carefully select the parent folder--even though it is obvious where the user wants to upload to.

This is interesting, but I think it's too prone to confusion as others have stated.